### PR TITLE
Refactoring the Project

### PR DESCRIPTION
**Pull request by DanielEJones**
Sorry to drop this massive PR on you guys. Sleep-deprived me though it would be a great idea to move everything around in one giant commit.

# What
Thankfully although this PR has a large amount of files changed, most of it is to do with formatting. The changes, in order of importance, are as follows:

1. Implemented the basic database model, adding charts and files alongside users
2. Moved the database over to use Migration for maintaining consistent models and database initialization
3. Changed `__init__.py` to contain an app factory rather than the app itself, which will make testing easier
4. Added new `config.py` and `main.py` files to hold config objects and make use of the factory respectively
5. Changed `routes.py` to use a routes blueprint rather than referring to `app` to support the above changes
6. Moved around a bunch of functions in `routes.py` and added section headers to help clarify where things are found
7. Split out `models` into a package to make it easier to read than one big file (may do the same to `routes` eventually)
8. Renamed `logic/` to `services/` and split the files into smaller ones (I'm not sold on the name, may go back to logic?)
9. Added more tests for the `services/` module and fixed some edge cases with the plot functions / registry
10. Updated `README` and `.gitignore` to reflect state of project
11. Fixed a few assorted typos

# Why
Hopefully a cleaner workspace will be helpful in the final push to implement functionality, as well as make it easier for the markers to navigate the code base.

### CODE REVIEW

**Comment by DanielEJones** [comment]
> flask --app app:create_app db init

To initialise the database you can just run `flask db init`

Further instructions can be found in the [updated README](https://github.com/oMakaron/AgileWebDev88/blob/refactor/README.md) for updating the database or running the program. Let me know if there is anything that is unclear and I'll try and clear it up.

**Comment by yuki-kii** [comment]
when im doing flask db init, it still say that
Error: Detected factory 'create_app' in module 'app', but could not call it without arguments. Use 'app:create_app(args)' to specify arguments.

**Comment by yuki-kii** [comment]
flask --app "app:create_app('config.DeploymentConfig')" db migrate -m "whatever"
flask --app "app:create_app('config.DeploymentConfig')" db upgrade
flask --app "app:create_app('config.DeploymentConfig')" run
im trying to do this, then it can be run

**Comment by yuki-kii** [comment]
when im trying to login it said Internal Server Error: RuntimeError: A secret key is required to use CSRF.

**Comment by DanielEJones** [comment]
> when im doing flask db init, it still say that

So I think I've worked out what it was:
1. Flask didn't know that it should be using `main.py` as the app, so we need to set the `FLASK_APP` environment variable
2. I don't think the `db init` command is the right one, it seems like calling upgrade works though.

To make it easier to use I've created a python script that will run all the necessary functions to set the project up. If you could give it a try and tell me if it works that would be great.

File: setup.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    print("Installing Tailwind...")
+    if not which("npm"):
+        print("npm is not installed. Please install it and rerun this script.")
+        return
+
+    check_call(["npm", "install", "-D", "tailwindcss@3"])
+
+    print("Building CSS...")
+    check_call([
+        "npx", "tailwindcss",
```
Comment (WillVetter) [inline setup.py:33]:
```suggestion
        npx_path, "tailwindcss",
```

File: setup.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        print("Virtual environment already exists.")
+
+    pip_exec = venv_path / ("Scripts" if platform == "win32" else "bin") / "pip"
+    flask_exec = venv_path / ("Scripts" if platform == "win32" else "bin") / "flask"
+
+    print("Installing dependencies...")
+    check_call([str(pip_exec), "install", "-r", "requirements.txt"])
+
+    print("Installing Tailwind...")
+    if not which("npm"):
```
Comment (WillVetter) [inline setup.py:25]:
```suggestion
    npx_path = which("npx")
    if not npx_path:
```

Comment (WillVetter) [reply setup.py:25]: Script was crashing on Windows because npx wasn’t being found by subprocess. This change fixes it by using which("npx") to grab the full path before running the command — works reliably across platforms now.

File: README.md [hunk 3]
```diff
@@ -41,19 +44,30 @@ The `output.css` file is ignored in the repository and must be generated locally
...
+```bash
+flask db migrate -m "migration message" # if you are the one who made changes
+flask db upgrade
+```
 
+### 6. Running the Application
+In order for the application to run, you will need to have a secret key set within the environment. You can set one using the following if you didn't set one with the install script:
 ```bash
-python init_database.py
+export SECRET_KEY='your-key-here'
```
Comment (WillVetter) [inline README.md:65]:
```suggestion
export FLASK_SECRET_KEY='your-key-here'
```

Comment (WillVetter) [reply README.md:65]: should be FLASK_SECRET_KEY instead of SECRET_KEY so Flask actually picks it up.

**Comment by yuki-kii** [comment]
need to change the readme for tailwind, because the input file for css is in the ./app/static/css/input.css.

**Review by WillVetter — COMMENTED** [review]
Nice job on this PR. The refactor with blueprints and the updated README make everything cleaner and easier to follow. Just a few small just adjustments but overall looks good to me.

File: README.md (hunk 3 above)
Comment (DanielEJones) [reply README.md:65]: Good catch

File: setup.py (hunk 2 above)
Comment (DanielEJones) [reply setup.py]: Good idea, I've changed it so both `npm` and `npx` are located using `which`.

**Review by yuki-kii — APPROVED** [review]
All good to me.

**Review by WillVetter — COMMENTED** [review]
I tested the setup.py file, and I think the issue is that create_app() requires a config argument, and Flask isn't able to call it automatically when running commands like flask db upgrade. The CLI doesn't know how to construct the app without that argument.

To fix it, I added a cli_app() function that returns create_app(DeploymentConfig), and updated the setup script to use --app=app:cli_app so that Flask could run the database commands properly.

**Review by oMakaron — APPROVED** [review]

File: .gitignore [hunk 4]
```diff
@@ -22,15 +21,16 @@ app/users.db
 dataFrame.csv
```
[the remark this answers is not in the document]
Comment (DanielEJones) [reply .gitignore:21]: Good point, I'll take that out

File: app/routes.py [hunk 5]
```diff
@@ -82,45 +73,85 @@ def signup():
...
-    return render_template('profile.html', user=user)
+def friends():
+    return render_template('friends.html')
 
-@app.route('/add-friend', methods=['GET', 'POST'])
+
+@bp.route('/add-friend', methods=['GET', 'POST'])
 @login_required
 def add_friend():
     return render_template('add_friend.html')
```
[the remark this answers is not in the document]
Comment (DanielEJones) [reply app/routes.py:147]: Good point, that might be a separate issue you could create
